c-ares module: missing symbols

The BUILD file:
```bazel
cc_binary (
  name = "async_dns",
  srcs = [ "async_dns.cc" ],
  deps = [ "@c-ares//:ares" ]
)
```

The C++ source file: async_dns.cc:
```c++
#include <ares.h>
#include <arpa/inet.h>
#include <ctype.h>
#include <netdb.h>
#include <netinet/in.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <unistd.h>

static void state_cb(void *data, int s, int read, int write) {
  printf("Change state fd %d read:%d write:%d\n", s, read, write);
}

static void callback(void *arg, int status, int timeouts,
                     struct hostent *host) {
  if (!host || status != ARES_SUCCESS) {
    printf("Failed to lookup %s\n", ares_strerror(status));
    return;
  }

  printf("Found address name %s\n", host->h_name);
  char ip[INET6_ADDRSTRLEN];
  int i = 0;

  for (i = 0; host->h_addr_list[i]; ++i) {
    inet_ntop(host->h_addrtype, host->h_addr_list[i], ip, sizeof(ip));
    printf("%s\n", ip);
  }
}

static void wait_ares(ares_channel channel) {
  for (;;) {
    struct timeval *tvp, tv;
    fd_set read_fds, write_fds;
    int nfds;

    FD_ZERO(&read_fds);
    FD_ZERO(&write_fds);
    nfds = ares_fds(channel, &read_fds, &write_fds);
    if (nfds == 0) {
      break;
    }
    tvp = ares_timeout(channel, NULL, &tv);
    select(nfds, &read_fds, &write_fds, NULL, tvp);
    ares_process(channel, &read_fds, &write_fds);
  }
}

int main(void) {
  ares_channel channel;
  int status;
  struct ares_options options;
  int optmask = 0;

  status = ares_library_init(ARES_LIB_INIT_ALL);
  if (status != ARES_SUCCESS) {
    printf("ares_library_init: %s\n", ares_strerror(status));
    return 1;
  }
  // options.sock_state_cb_data;
  options.sock_state_cb = state_cb;
  optmask |= ARES_OPT_SOCK_STATE_CB;

  status = ares_init_options(&channel, &options, optmask);
  if (status != ARES_SUCCESS) {
    printf("ares_init_options: %s\n", ares_strerror(status));
    return 1;
  }

  status = ares_set_servers_csv(channel, "114.114.114.114");
  if (status != ARES_SUCCESS) {
    printf("ares_set_servers_csv: %s\n", ares_strerror(status));
    return 1;
  }

  ares_gethostbyname(channel, "baidu.com", AF_INET, callback, NULL);
  // ares_gethostbyname(channel, "google.com", AF_INET6, callback, NULL);
  wait_ares(channel);
  ares_destroy(channel);
  ares_library_cleanup();
  printf("fin\n");
  return 0;
}
```

Use Bazel to build the C++ source file:
```
$ bazelisk build -s :async_dns
INFO: Analyzed target //:async_dns (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
SUBCOMMAND: # //:async_dns [action 'Linking async_dns', configuration: 386057a8d24831c33aaf953f87aec3b0ebdbf5a779460907e5f2832f79a9e67a, execution platform: @local_config_platform//:host]
(cd /private/var/tmp/_bazel_jing/352d1d854f060ba5704eda8af920a5e3/execroot/__main__ && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=11.1 \
    PATH='/Users/jing/Library/Caches/bazelisk/downloads/bazelbuild/bazel-5.1.0-darwin-x86_64/bin:/usr/local/opt/node@16/bin:/Users/jing/.rbenv/shims:/Users/jing/.tiup/bin:/Users/jing/opt/anaconda3/condabin:/usr/local/sbin:/Users/jing/code/github/zk-code/scripts:/Users/jing/tools/apache-zookeeper-3.5.5-bin/bin:/Users/jing/code/github/trace/sky/skywalking-cli/bin:/Users/jing/tools/apache-skywalking-apm-bin-es7/bin:/usr/local/opt/ruby/bin:/Users/jing/code/github/devops/fluent-bit/build/bin:/Users/jing/tools/apache-maven-3.6.3/bin:/Users/jing/.gem/bin:/Users/jing/.cargo/bin:/Users/jing/code/github/network/caddy/cmd/caddy:/usr/local/opt/coreutils/libexec/gnubin:/Users/jing/tools/apache-skywalking-apm-bin-es7/bin:/usr/local/opt/bison/bin:/usr/local/opt/findutils/libexec/gnubin:/usr/local/opt/gnu-sed/libexec/gnubin:/Users/jing/tools/mongodb-osx-x86_64-4.0.2/bin:/usr/local/opt/mysql@5.7/bin:/Users/jing/tools/spark-2.2.0-bin-hadoop2.7/bin:/Users/jing/bin:/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/Library/Apple/usr/bin:/Applications/Wireshark.app/Contents/MacOS:/Users/jing/go/bin:/Users/jing/.pub-cache/bin:/Applications/Visual Studio Code.app/Contents/Resources/app/bin:/usr/local/opt/fzf/bin' \
    XCODE_VERSION_OVERRIDE=12.4.0.12D4e \
    ZERO_AR_DATE=1 \
  external/local_config_cc/cc_wrapper.sh @bazel-out/darwin-fastbuild/bin/async_dns-2.params)
# Configuration: 386057a8d24831c33aaf953f87aec3b0ebdbf5a779460907e5f2832f79a9e67a
# Execution platform: @local_config_platform//:host
ERROR: /Users/jing/code/xdf/sac/dns-prober/BUILD.bazel:47:11: Linking async_dns failed: (Aborted): cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh @bazel-out/darwin-fastbuild/bin/async_dns-2.params

Use --sandbox_debug to see verbose messages from the sandbox
Undefined symbols for architecture x86_64:
  "_ares__freeaddrinfo_cnames", referenced from:
      _ares_parse_a_reply in libares.lo(ares_parse_a_reply.o)
      _ares_parse_aaaa_reply in libares.lo(ares_parse_aaaa_reply.o)
  "_ares__freeaddrinfo_nodes", referenced from:
      _ares_parse_a_reply in libares.lo(ares_parse_a_reply.o)
      _ares_parse_aaaa_reply in libares.lo(ares_parse_aaaa_reply.o)
  "_ares__parse_into_addrinfo2", referenced from:
      _ares_parse_a_reply in libares.lo(ares_parse_a_reply.o)
      _ares_parse_aaaa_reply in libares.lo(ares_parse_aaaa_reply.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error in child process '/usr/bin/xcrun'. 1
external/local_config_cc/cc_wrapper.sh: line 69: 72006 Abort trap: 6           "$(/usr/bin/dirname "$0")"/wrapped_clang "$@"
Target //:async_dns failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.350s, Critical Path: 0.11s
INFO: 2 processes: 2 internal.
FAILED: Build did NOT complete successfully
```
The build shows that some symbols such as `_ares__freeaddrinfo_nodes` are missing. And a check of `libares.lo` confirms that `_ares__freeaddrinfo_nodes` is undefined:

```
$ cat bazel-out/darwin-fastbuild/bin/async_dns-2.params
-lc++
-fobjc-link-runtime
-Wl,-S
-o
bazel-out/darwin-fastbuild/bin/async_dns
bazel-out/darwin-fastbuild/bin/_objs/async_dns/async_dns.o
-Wl,-force_load,bazel-out/darwin-fastbuild/bin/external/c-ares.1.16.1/libares.lo
-headerpad_max_install_names
-no-canonical-prefixes
-target
x86_64-apple-macosx
-mmacosx-version-min=11.1
-lc++
-target
x86_64-apple-macosx

$ nm -g bazel-out/darwin-fastbuild/bin/external/c-ares.1.16.1/libares.lo | ag ares__freeaddrinfo_nodes
no symbols
no symbols
no symbols
no symbols
no symbols
no symbols
                 U _ares__freeaddrinfo_nodes
                 U _ares__freeaddrinfo_nodes
```

`ares__freeaddrinfo_nodes` is contained in https://github.com/c-ares/c-ares/blob/cares-1_16_1/ares_freeaddrinfo.c. But libares.lo does not include it.
```
$ ar -t bazel-out/darwin-fastbuild/bin/external/c-ares.1.16.1/libares.lo
__.SYMDEF SORTED
ares__close_sockets.o
ares__get_hostent.o
ares__read_line.o
ares__timeval.o
ares_android.o
ares_cancel.o
ares_create_query.o
ares_data.o
ares_destroy.o
ares_expand_name.o
ares_expand_string.o
ares_fds.o
ares_free_hostent.o
ares_free_string.o
ares_getenv.o
ares_gethostbyaddr.o
ares_gethostbyname.o
ares_getnameinfo.o
ares_getopt.o
ares_getsock.o
ares_init.o
ares_library_init.o
ares_llist.o
ares_mkquery.o
ares_nowarn.o
ares_options.o
ares_parse_a_reply.o
ares_parse_aaaa_reply.o
ares_parse_mx_reply.o
ares_parse_naptr_reply.o
ares_parse_ns_reply.o
ares_parse_ptr_reply.o
ares_parse_soa_reply.o
ares_parse_srv_reply.o
ares_parse_txt_reply.o
ares_platform.o
ares_process.o
ares_query.o
ares_search.o
ares_send.o
ares_strcasecmp.o
ares_strdup.o
ares_strsplit.o
ares_strerror.o
ares_timeout.o
ares_version.o
ares_writev.o
bitncmp.o
inet_net_pton.o
inet_ntop.o
windows_port.o
```
The BUILD file for c-ares module does not include `ares_freeaddrinfo.c`:
https://github.com/bazelbuild/bazel-central-registry/blob/53e8c3ad84c9407df5d644fef680fe92f27e7a39/modules/c-ares/1.16.1/patches/add_build_file.patch#L67-L119

--- a/modules/c-ares/1.16.1/patches/add_build_file.patch
+++ b/modules/c-ares/1.16.1/patches/add_build_file.patch
@@ -1,6 +1,6 @@
 --- /dev/null
 +++ BUILD.bazel
-@@ -0,0 +1,170 @@
+@@ -0,0 +1,173 @@
 +load("@bazel_skylib//rules:copy_file.bzl", "copy_file")
 +
 +config_setting(
@@ -67,7 +67,10 @@
 +    srcs = [
 +        "ares__close_sockets.c",
 +        "ares__get_hostent.c",
++        "ares__parse_into_addrinfo.c",
 +        "ares__read_line.c",
++        "ares__readaddrinfo.c",
++        "ares__sortaddrinfo.c",
 +        "ares__timeval.c",
 +        "ares_android.c",
 +        "ares_cancel.c",
@@ -79,11 +82,12 @@
 +        "ares_fds.c",
 +        "ares_free_hostent.c",
 +        "ares_free_string.c",
++        "ares_freeaddrinfo.c",
++        "ares_getaddrinfo.c",
 +        "ares_getenv.c",
 +        "ares_gethostbyaddr.c",
 +        "ares_gethostbyname.c",
 +        "ares_getnameinfo.c",
-+        "ares_getopt.c",
 +        "ares_getsock.c",
 +        "ares_init.c",
 +        "ares_library_init.c",
@@ -125,7 +129,6 @@
 +        "ares_data.h",
 +        "ares_dns.h",
 +        "ares_getenv.h",
-+        "ares_getopt.h",
 +        "ares_inet_net_pton.h",
 +        "ares_iphlpapi.h",
 +        "ares_ipv6.h",

--- a/modules/c-ares/1.16.1/source.json
+++ b/modules/c-ares/1.16.1/source.json
@@ -2,7 +2,7 @@
     "integrity": "sha256-0IMS0OzDvUju4KTMDSE3yfGU4KKN4gKJKMD2yuhfhs4=",
     "patch_strip": 0,
     "patches": {
-        "add_build_file.patch": "sha256-+SUCFxBIkR0GE9FRFPps/e6AnA9cQIGANBHK14UAKwQ="
+        "add_build_file.patch": "sha256-BfqAaVGKiHH4Cg0dPisG+ycnHKlogIgO1znwG8tNIVY="
     },
     "strip_prefix": "c-ares-1.16.1",
     "url": "https://github.com/c-ares/c-ares/releases/download/cares-1_16_1/c-ares-1.16.1.tar.gz"

